skyportal/tests/api/test_bokeh_plots.py

```diff
@@ -0,0 +1,177 @@
+import os
+from skyportal.tests import api
+import numpy as np
+import datetime
+from bokeh.util import serialization
+
+
+def black_body(la, temp):
+    """
+    get the amount of radiation expected from a black body
+    of the given temperature "temp", at the given wavelengths "la".
+    :param la: float array or scalar
+        wavelength(s) where the radiation should be calculated.
+    :param temp: float scalar
+        temperature of the black body.
+    :return:
+        return the radiation (in units of Watts per steradian per m^2 per nm)
+    """
+    const = (
+        0.014387773538277204  # h*c/k_b = 6.62607004e-34 * 299792458 / 1.38064852e-23
+    )
+    amp = 1.1910429526245744e-25  # 2*h*c**2 * (nm / m) = 2*6.62607004e-34 * 299792458**2 / 1e9 the last term is units
+    la = la * 1e-9  # convert wavelength from nm to m
+
+    return amp / (la ** 5 * (np.exp(const / (la * temp)) - 1))
+
+
+def get_plot_data(data):
+    """
+    Parse the dictionary "data" received from a GET request
+    to api/internal/plot/spectroscopy/<obj_id> and recover
+    the underlying data for all spectra in the plot.
+
+    Returns a list of dictionaries with the flux, wavelength, etc.
+    """
+    objects = []
+    for doc in data['data']['bokehJSON']['doc']['roots']['references']:
+        if 'data' in doc['attributes']:
+            new_obj = {}
+            for field_name in ['wavelength', 'flux', 'flux_original', 'x', 'y']:
+                if field_name in doc['attributes']['data']:
```

Comment:
Maybe this should be refactored into a small utility function, or an inline function?  It's quite hard to read what's happening here.

Reply:
Yeah, there's JSONPath, but I don't think that really helps matters along much.

Easiest is to just assign meaningful names, e.g.:

```
attributes = doc['attributes']['data']
if field_name in attributes:
    ...
```

I.e., try and get it to read as close to English.  But, I hear you.